File: windows/tests/ConnectorControl.Core.Tests/State/AppStateWatcherTests.cs
using ConnectorControl.Core.Services;
using ConnectorControl.Core.State;
using ConnectorControl.Core.Tests.TestSupport;

namespace ConnectorControl.Core.Tests.State;

public class AppStateWatcherTests
{
    private static readonly TimeSpan Wait = TimeSpan.FromSeconds(8);
    private static readonly TimeSpan Settle = TimeSpan.FromMilliseconds(1500);
    private static readonly string[] Fixture = ["aws-mcp", "scoutbook", "service-now"];

    [Fact]
    public void ClaudeConfigWatcherRegeneratesAnExternalEdit()
    {
        using var h = new AppStateHarness();
        using var state = h.Create();
        Thread.Sleep(300);
        h.WriteClaudeServers(("scoutbook", state.Store.Mcps["scoutbook"].Config));
        Assert.True(h.Ui.PumpUntil(() => h.Notifier.Sent.Count == 1, Wait));
        Assert.Equal(AppState.ClaudeConfigRegeneratedBody, h.Notifier.Sent[0].Body);
        Assert.Equal(Fixture, AppStateHarness.Keys(h.ClaudeServers().Keys));
        h.Ui.PumpUntil(() => false, Settle);   // the regenerating write echoes through the watcher: it must stay quiet
        Assert.Single(h.Notifier.Sent);
    }

    [Fact]
    public void StoreWatcherAdoptsAnExternalStoreAndAnnouncesTheRestart()
    {
        using var h = new AppStateHarness();
        h.Claude.IsRunning = true;
        h.Claude.LaunchTime = h.Now.AddHours(-1);
        using var state = h.Create();
        Thread.Sleep(300);
        var synced = h.StoreOnDisk();
        synced.Mcps["scoutbook"] = synced.Mcps["scoutbook"] with { Enabled = false };
        MasterStoreIO.Save(synced, h.MasterStorePath);   // another machine's list arrives via sync
        Assert.True(h.Ui.PumpUntil(() => h.Notifier.Sent.Count == 1, Wait));
        Assert.Equal((Notifications.Title, AppState.ConnectorListChangedRestartBody, (string?)Notifications.RestartCategory), h.Notifier.Sent[0]);
        Assert.False(state.Store.Mcps["scoutbook"].Enabled);
        Assert.Equal(["aws-mcp", "service-now"], AppStateHarness.Keys(h.ClaudeServers().Keys));
        Assert.True(state.NeedsClaudeRestart);
    }

    [Fact]
    public void StoreWatcherIgnoresOurOwnWriteEcho()
    {
        using var h = new AppStateHarness();
        using var state = h.Create();
        Thread.Sleep(300);
        state.SetEnabled("aws-mcp", false);
        h.Ui.PumpUntil(() => false, Settle);
        Assert.Empty(h.Notifier.Sent);
        Assert.False(state.Store.Mcps["aws-mcp"].Enabled);
    }

    [Fact]
    public void StoreWatcherIgnoresAnUndecodablePartialWrite()
    {
        using var h = new AppStateHarness();
        using var state = h.Create();
        Thread.Sleep(300);
        File.WriteAllText(h.MasterStorePath, "{\"version\": 2, \"acti");   // a sync tool mid-write
        h.Ui.PumpUntil(() => false, Settle);
        Assert.Equal(Fixture, state.SortedNames);
        Assert.Equal("{\"version\": 2, \"acti", File.ReadAllText(h.MasterStorePath));   // not moved aside: no reload happened
        Assert.Empty(h.Notifier.Sent);
    }

    [Fact]
    public void DeletedStoreFileIsRePersistedFromMemory()
    {
        using var h = new AppStateHarness();
        using var state = h.Create();
        Thread.Sleep(300);
        File.Delete(h.MasterStorePath);
        Assert.True(h.Ui.PumpUntil(() => File.Exists(h.MasterStorePath), Wait));
        Assert.Equal(state.Store, h.StoreOnDisk());
        Assert.Empty(h.Notifier.Sent);
    }

    [Fact]
    public void ReloadArmsOnlyTheWatcherThatCouldNotArmYet()
    {
        using var h = new AppStateHarness();
        using var state = h.Create();
        Assert.True(state.WatchersArmed);
        var later = h.Dir.File(Path.Combine("later", "claude_desktop_config.json"));
        Assert.False(Directory.Exists(Path.GetDirectoryName(later)!));

        // ArmWatchers cannot arm a watcher whose parent directory is missing; the
        // regenerating write inside the same Reload creates it, and the re-arm at the
        // end of Reload catches up (spec §6.3).
        state.RepointClaudeConfig(later);
        Assert.True(File.Exists(later));
        Assert.True(state.WatchersArmed);

        Thread.Sleep(300);
        ClaudeConfigIO.Write(new Dictionary<string, JsonValue> { ["only"] = AppStateHarness.Remote("https://only.example/mcp") }, later);
        Assert.True(h.Ui.PumpUntil(() => h.Notifier.Sent.Count == 1, Wait));   // the new location really is watched
        Assert.Equal(AppState.ClaudeConfigRegeneratedBody, h.Notifier.Sent[0].Body);
    }

    [Fact]
    public void RepointStoreSeedsAnEmptyLocationAndReArmsTheWatcher()
    {
        using var h = new AppStateHarness();
        using var state = h.Create();
        var synced = h.Dir.File("synced");
        state.RepointStore(synced);
        Assert.Equal(synced, h.Settings.MasterStoreDir);
        Assert.Equal(synced, state.Service.Paths.StoreDir);
        Assert.Equal(h.BackupsDir, state.Service.Paths.BackupsDir);   // backups never follow the store
        Assert.Equal(state.Store, MasterStoreIO.Read(Path.Combine(synced, "mcps.json")));
        Assert.True(File.Exists(h.MasterStorePath));   // the previous file is never deleted
        Assert.Empty(h.Notifier.Sent);

        Thread.Sleep(300);
        var edited = state.Store.Clone();
        edited.Mcps["aws-mcp"] = edited.Mcps["aws-mcp"] with { Enabled = false };
        MasterStoreIO.Save(edited, Path.Combine(synced, "mcps.json"));
        Assert.True(h.Ui.PumpUntil(() => !state.Store.Mcps["aws-mcp"].Enabled, Wait));   // the new location is watched
    }

    [Fact]
    public void RepointStoreAdoptsAnExistingStoreQuietly()
    {
        using var h = new AppStateHarness();
        using var state = h.Create();
        var synced = h.Dir.File("synced");
        Directory.CreateDirectory(synced);
        var theirs = state.Store.Clone();
        theirs.Mcps["aws-mcp"] = theirs.Mcps["aws-mcp"] with { Enabled = false };
        theirs.Mcps["synced-only"] = new McpEntry(AppStateHarness.Remote("https://synced.example/mcp"));
        MasterStoreIO.Save(theirs, Path.Combine(synced, "mcps.json"));

        state.RepointStore(synced);
        Assert.Equal(["aws-mcp", "scoutbook", "service-now", "synced-only"], state.SortedNames);
        Assert.False(state.Store.Mcps["aws-mcp"].Enabled);
        Assert.Equal(["scoutbook", "service-now", "synced-only"], AppStateHarness.Keys(h.ClaudeServers().Keys));   // regenerated
        Assert.Empty(h.Notifier.Sent);   // quiet adoption: the user is watching
    }

    [Fact]
    public void RepointStoreBackToTheDefault()
    {
        using var h = new AppStateHarness();
        using var state = h.Create();
        state.RepointStore(h.Dir.File("synced"));
        state.RepointStore(null);
        Assert.Null(h.Settings.MasterStoreDir);
        Assert.Equal(h.StoreDir, state.Service.Paths.StoreDir);
        Assert.Equal(Fixture, state.SortedNames);
    }

    [Fact]
    public void RefreshServiceSettingsAppliesTheKeepCountWithoutReloading()
    {
        using var h = new AppStateHarness();
        using var state = h.Create();
        h.Settings.BackupKeepCount = 7;
        state.RefreshServiceSettings();
        Assert.Equal(7, state.Service.Backups.KeepCount);
        Assert.Null(h.Settings.LastApplyDate);
        Assert.Equal(Fixture, state.SortedNames);
    }

    [Fact]
    public void RepointClaudeConfigImportsTheNewFileFreshlyAndQuietly()
    {
        using var h = new AppStateHarness();
        using var state = h.Create();
        var other = h.Dir.File(Path.Combine("other", "claude_desktop_config.json"));
        Directory.CreateDirectory(Path.GetDirectoryName(other)!);
        ClaudeConfigIO.Write(new Dictionary<string, JsonValue> { ["other-only"] = AppStateHarness.Remote("https://other.example/mcp") }, other);

        state.RepointClaudeConfig(other);
        Assert.Equal(other, h.Settings.ClaudeConfigPath);
        Assert.Equal(other, state.Service.Paths.ClaudeConfigPath);
        Assert.Equal(["aws-mcp", "other-only", "scoutbook", "service-now"], state.SortedNames);
        Assert.Equal(["aws-mcp", "other-only", "scoutbook", "service-now"], AppStateHarness.Keys(ClaudeConfigIO.ReadMcpServers(other).Keys));
        Assert.Equal(Fixture, AppStateHarness.Keys(h.ClaudeServers().Keys));   // the old file is left alone
        Assert.Empty(h.Notifier.Sent);

        state.RepointClaudeConfig(null);
        Assert.Null(h.Settings.ClaudeConfigPath);
        Assert.Equal(h.ClaudeConfigPath, state.Service.Paths.ClaudeConfigPath);
    }

    [Fact]
    public void RestoreClaudeConfigIsQuietAndSyncsTheBaseline()
    {
        using var h = new AppStateHarness();
        using var state = h.Create();
        state.SetEnabled("aws-mcp", false);
        h.Notifier.Sent.Clear();
        h.Now = h.Now.AddMinutes(5);
        var backup = state.Service.Backups.Backups("claude_desktop_config")[0];   // the three-server file from before the toggle

        state.RestoreClaudeConfig(backup);
        Assert.Equal(Fixture, AppStateHarness.Keys(h.ClaudeServers().Keys));
        Assert.True(state.Store.Mcps["aws-mcp"].Enabled);   // adopted into the store
        Assert.Equal(Fixture, AppStateHarness.Keys(state.AppliedServers.Keys));
        Assert.Equal(h.Now, h.Settings.LastApplyDate);
        Assert.False(state.IsDirty);
        Assert.Empty(h.Notifier.Sent);
    }

    [Fact]
    public void RestoreFailurePropagatesWithoutTouchingTheFile()
    {
        using var h = new AppStateHarness();
        using var state = h.Create();
        var bad = h.Dir.File("bad.json");
        File.WriteAllText(bad, "{not json");
        var before = File.ReadAllBytes(h.ClaudeConfigPath);
        var ex = Assert.Throws<ClaudeConfigException>(() => state.RestoreClaudeConfig(bad));
        Assert.Equal("backup bad.json is not a valid config file", ex.Detail);
        Assert.Equal(before, File.ReadAllBytes(h.ClaudeConfigPath));
    }

    [Fact]
    public void DisposeStopsTheWatchers()
    {
        using var h = new AppStateHarness();
        var state = h.Create();
        Thread.Sleep(300);
        state.Dispose();
        h.WriteClaudeServers(("scoutbook", state.Store.Mcps["scoutbook"].Config));
        h.Ui.PumpUntil(() => false, Settle);
        Assert.Empty(h.Notifier.Sent);
        Assert.Equal(["scoutbook"], AppStateHarness.Keys(h.ClaudeServers().Keys));   // nobody regenerated it
    }
}

File: Tests/ConnectorControlStateTests/AppStateWatcherTests.swift
import XCTest
import ConnectorControlCore
@testable import ConnectorControlState

/// windows/tests/ConnectorControl.Core.Tests/State/AppStateWatcherTests.cs
/// against the real DispatchSource watchers. The 300 ms sleeps keep the
/// external write's mtime distinct from the file the harness just wrote.
@MainActor
final class AppStateWatcherTests: XCTestCase {
    private let wait: TimeInterval = 8
    private let settle: TimeInterval = 1.5
    private let fixture = ["aws-mcp", "scoutbook", "service-now"]

    func testClaudeConfigWatcherRegeneratesAnExternalEdit() throws {
        let h = AppStateHarness()
        defer { h.dispose() }
        let state = h.create()
        Thread.sleep(forTimeInterval: 0.3)
        try h.writeClaudeServers([("scoutbook", try XCTUnwrap(state.store.mcps["scoutbook"]).config)])
        XCTAssertTrue(h.ui.pumpUntil({ h.notifier.sent.count == 1 }, timeout: wait))
        XCTAssertEqual(h.notifier.sent[0].body, AppState.claudeConfigRegeneratedBody)
        XCTAssertEqual(try h.claudeServers().keys.sorted(), fixture)
        _ = h.ui.pumpUntil({ false }, timeout: settle)   // the regenerating write echoes through the watcher: it must stay quiet
        XCTAssertEqual(h.notifier.sent.count, 1)
    }

    func testExternalEditIsSilentWhenNotifyExternalChangesIsDisabledButStillReloads() throws {
        let h = AppStateHarness()
        defer { h.dispose() }
        h.settings.notifyExternalChanges = false
        let state = h.create()
        Thread.sleep(forTimeInterval: 0.3)
        try h.writeClaudeServers([("scoutbook", try XCTUnwrap(state.store.mcps["scoutbook"]).config)])
        XCTAssertTrue(h.ui.pumpUntil({ (try? h.claudeServers().keys.sorted()) == self.fixture }, timeout: wait))
        _ = h.ui.pumpUntil({ false }, timeout: settle)   // give the regenerating write's own echo a chance to fire too
        XCTAssertTrue(h.notifier.sent.isEmpty)
    }

    func testStoreWatcherAdoptsAnExternalStoreAndAnnouncesTheRestart() throws {
        let h = AppStateHarness()
        defer { h.dispose() }
        h.claude.isRunning = true
        h.claude.launchDate = h.now.addingTimeInterval(-3600)
        let state = h.create()
        Thread.sleep(forTimeInterval: 0.3)
        var synced = try h.storeOnDisk()
        synced.mcps["scoutbook"]?.enabled = false
        try MasterStoreIO.save(synced, to: h.masterStoreURL)   // another machine's list arrives via sync
        XCTAssertTrue(h.ui.pumpUntil({ h.notifier.sent.count == 1 }, timeout: wait))
        XCTAssertEqual(h.notifier.sent[0], FakeNotifier.Sent(
            title: Notifications.title,
            body: AppState.connectorListChangedBody(ServerDelta(removed: ["scoutbook"]), restartRequired: true),
            category: Notifications.restartCategory))
        XCTAssertEqual(state.store.mcps["scoutbook"]?.enabled, false)
        XCTAssertEqual(try h.claudeServers().keys.sorted(), ["aws-mcp", "service-now"])
        XCTAssertTrue(state.needsClaudeRestart)
    }

    /// A synced list can add a command Claude will run. With Claude not running there is no
    /// restart to offer, but the adoption still has to be announced — before, every branch of
    /// the notification chain missed this case and the new server simply started next launch.
    func testStoreWatcherAnnouncesAnAdoptedAdditionEvenWhenClaudeIsNotRunning() throws {
        let h = AppStateHarness()
        defer { h.dispose() }
        h.claude.isRunning = false
        let state = h.create()
        Thread.sleep(forTimeInterval: 0.3)
        var synced = try h.storeOnDisk()
        synced.mcps["evil"] = MCPEntry(config: .object(["command": .string("curl"), "args": .array([.string("https://x.example/run")])]))
        try MasterStoreIO.save(synced, to: h.masterStoreURL)   // another machine's list arrives via sync
        XCTAssertTrue(h.ui.pumpUntil({ h.notifier.sent.count == 1 }, timeout: wait))
        XCTAssertEqual(h.notifier.sent[0], FakeNotifier.Sent(
            title: Notifications.title,
            body: AppState.connectorListChangedBody(ServerDelta(added: ["evil"]), restartRequired: false),
            category: nil))
        XCTAssertEqual(try h.claudeServers().keys.sorted(), ["aws-mcp", "evil", "scoutbook", "service-now"])
        XCTAssertFalse(state.needsClaudeRestart)
    }

    func testStoreWatcherIgnoresOurOwnWriteEcho() {
        let h = AppStateHarness()
        defer { h.dispose() }
        let state = h.create()
        Thread.sleep(forTimeInterval: 0.3)
        state.setEnabled("aws-mcp", false)
        _ = h.ui.pumpUntil({ false }, timeout: settle)
        XCTAssertTrue(h.notifier.sent.isEmpty)
        XCTAssertEqual(state.store.mcps["aws-mcp"]?.enabled, false)
    }

    func testStoreWatcherIgnoresAnUndecodablePartialWrite() throws {
        let h = AppStateHarness()
        defer { h.dispose() }
        let state = h.create()
        Thread.sleep(forTimeInterval: 0.3)
        try Data("{\"version\": 2, \"acti".utf8).write(to: h.masterStoreURL)   // a sync tool mid-write
        _ = h.ui.pumpUntil({ false }, timeout: settle)
        XCTAssertEqual(state.sortedNames, fixture)
        XCTAssertEqual(try String(contentsOf: h.masterStoreURL, encoding: .utf8), "{\"version\": 2, \"acti")   // not moved aside: no reload happened
        XCTAssertTrue(h.notifier.sent.isEmpty)
    }

    func testDeletedStoreFileIsRePersistedFromMemory() throws {
        let h = AppStateHarness()
        defer { h.dispose() }
        let state = h.create()
        Thread.sleep(forTimeInterval: 0.3)
        try FileManager.default.removeItem(at: h.masterStoreURL)
        XCTAssertTrue(h.ui.pumpUntil({ FileManager.default.fileExists(atPath: h.masterStoreURL.path) }, timeout: wait))
        XCTAssertEqual(try h.storeOnDisk(), state.store)
        XCTAssertTrue(h.notifier.sent.isEmpty)
    }

    /// The C# version repoints the Claude config into a directory that does not
    /// exist yet; the Mac has no such repoint, so the missing directory is the
    /// Claude folder itself at launch (spec §6.3: both watchers live after every reload).
    func testReloadArmsOnlyTheWatcherThatCouldNotArmYet() throws {
        let h = AppStateHarness(seedClaudeConfig: false, createClaudeDirectory: false)
        defer { h.dispose() }
        let state = h.create()
        // No Claude folder and no store written (an empty reconcile saves nothing): neither watcher could arm.
        XCTAssertFalse(state.watchersArmed)
        XCTAssertNil(state.upsert(name: "only", entry: MCPEntry(config: AppStateHarness.remote("https://only.example/mcp")), renamedFrom: nil))
        state.applyInteractively()   // creates the Claude folder and the file
        XCTAssertTrue(FileManager.default.fileExists(atPath: h.claudeConfigURL.path))
        XCTAssertFalse(state.watchersArmed, "an apply is not a reload: nothing re-arms yet")
        state.reload()
        XCTAssertTrue(state.watchersArmed, "the re-arm at the end of reload caught up")
        // The "only" half: a further reload leaves the now-armed watchers alone
        // instead of tearing them down and re-baselining their mtimes.
        let armed = state.watcherIdentities
        state.reload()
        XCTAssertTrue(state.watchersArmed)
        XCTAssertEqual(state.watcherIdentities.claude, armed.claude, "an armed watcher is not replaced by reload")
        XCTAssertEqual(state.watcherIdentities.store, armed.store, "an armed watcher is not replaced by reload")

        Thread.sleep(forTimeInterval: 0.3)
        try h.writeClaudeServers([("only", AppStateHarness.remote("https://moved.example/mcp"))])
        XCTAssertTrue(h.ui.pumpUntil({ h.notifier.sent.count == 1 }, timeout: wait))   // the location really is watched
        XCTAssertEqual(h.notifier.sent[0].body, AppState.claudeConfigRegeneratedBody)
    }

    /// The store watcher's own directory disappearing (not just its file) is
    /// self-healing: the deletion's own reload re-persists the store from
    /// memory, recreating the directory the next reArm needs.
    func testDeletingTheStoreDirectoryReArmsOnTheNextReload() throws {
        let h = AppStateHarness()
        defer { h.dispose() }
        let state = h.create()
        XCTAssertTrue(state.watchersArmed)
        try FileManager.default.removeItem(at: h.storeDir)
        _ = h.ui.pumpUntil({ false }, timeout: settle)   // let the deletion's own detection run
        state.reload()
        XCTAssertTrue(state.watchersArmed)
    }

    func testRepointStoreSeedsAnEmptyLocationAndReArmsTheWatcher() throws {
        let h = AppStateHarness()
        defer { h.dispose() }
        let state = h.create()
        let synced = h.dir.file("synced")
        state.repointStore(to: synced)
        XCTAssertEqual(h.settings.masterStoreDir, synced.path)
        XCTAssertEqual(state.service.paths.storeDirURL.path, synced.path)
        XCTAssertEqual(state.service.paths.backupsDirURL.path, h.backupsDir.path)   // backups never follow the store
        XCTAssertEqual(MasterStoreIO.read(from: synced.appendingPathComponent("mcps.json")), state.store)
        XCTAssertTrue(FileManager.default.fileExists(atPath: h.masterStoreURL.path))   // the previous file is never deleted
        XCTAssertTrue(h.notifier.sent.isEmpty)

        Thread.sleep(forTimeInterval: 0.3)
        var edited = state.store
        edited.mcps["aws-mcp"]?.enabled = false
        try MasterStoreIO.save(edited, to: synced.appendingPathComponent("mcps.json"))
        XCTAssertTrue(h.ui.pumpUntil({ state.store.mcps["aws-mcp"]?.enabled == false }, timeout: wait))   // the new location is watched
    }

    func testRepointStoreAdoptsAnExistingStoreQuietly() throws {
        let h = AppStateHarness()
        defer { h.dispose() }
        let state = h.create()
        let synced = h.dir.file("synced")
        try FileManager.default.createDirectory(at: synced, withIntermediateDirectories: true)
        var theirs = state.store
        theirs.mcps["aws-mcp"]?.enabled = false
        theirs.mcps["synced-only"] = MCPEntry(config: AppStateHarness.remote("https://synced.example/mcp"))
        try MasterStoreIO.save(theirs, to: synced.appendingPathComponent("mcps.json"))

        state.repointStore(to: synced)
        XCTAssertEqual(state.sortedNames, ["aws-mcp", "scoutbook", "service-now", "synced-only"])
        XCTAssertEqual(state.store.mcps["aws-mcp"]?.enabled, false)
        XCTAssertEqual(try h.claudeServers().keys.sorted(), ["scoutbook", "service-now", "synced-only"])   // regenerated
        XCTAssertTrue(h.notifier.sent.isEmpty)   // quiet adoption: the user is watching
    }

    /// The new location exists but cannot be written into (its directory is
    /// write-blocked): the seed write throws, so the repoint must not adopt a
    /// service pointed at a store that was never actually seeded.
    func testARepointWhoseSeedCannotBeWrittenKeepsTheOldStore() throws {
        let h = AppStateHarness()
        defer { h.dispose() }
        let state = h.create()
        let target = h.dir.file("blocked")
        try FileManager.default.createDirectory(at: target, withIntermediateDirectories: true)
        let block = try WriteBlock(target.appendingPathComponent("mcps.json"))
        defer { block.dispose() }
        XCTAssertTrue(block.isEffective)
        let previousStoreDir = state.service.paths.storeDirURL
        let beforeClaudeConfig = try Data(contentsOf: h.claudeConfigURL)

        state.repointStore(to: target)

        XCTAssertEqual(state.service.paths.storeDirURL, previousStoreDir)
        XCTAssertNotNil(state.lastError)
        XCTAssertEqual(try Data(contentsOf: h.claudeConfigURL), beforeClaudeConfig)
        XCTAssertNil(h.settings.masterStoreDir, "restored to its previous value")
    }

    func testRepointStoreBackToTheDefault() {
        let h = AppStateHarness()
        defer { h.dispose() }
        let state = h.create()
        state.repointStore(to: h.dir.file("synced"))
        state.repointStore(to: nil)
        XCTAssertNil(h.settings.masterStoreDir)
        XCTAssertEqual(state.service.paths.storeDirURL.path, h.storeDir.path)
        XCTAssertEqual(state.sortedNames, fixture)
    }

    func testRefreshServiceSettingsAppliesTheKeepCountWithoutReloading() {
        let h = AppStateHarness()
        defer { h.dispose() }
        let state = h.create()
        h.settings.backupKeepCount = 7
        state.refreshServiceSettings()
        XCTAssertEqual(state.service.backups.keepCount, 7)
        XCTAssertNil(h.settings.lastApplyDate)
        XCTAssertEqual(state.sortedNames, fixture)
    }

    func testRestoreClaudeConfigIsQuietAndSyncsTheBaseline() throws {
        let h = AppStateHarness()
        defer { h.dispose() }
        let state = h.create()
        state.setEnabled("aws-mcp", false)
        h.notifier.clearSent()
        h.now = h.now.addingTimeInterval(300)
        let backup = try XCTUnwrap(state.service.backups.backups(series: "claude_desktop_config").first)   // the three-server file from before the toggle

        try state.restoreClaudeConfig(from: backup)
        XCTAssertEqual(try h.claudeServers().keys.sorted(), fixture)
        XCTAssertEqual(state.store.mcps["aws-mcp"]?.enabled, true)   // adopted into the store
        XCTAssertEqual(state.appliedServers.keys.sorted(), fixture)
        XCTAssertEqual(h.settings.lastApplyDate, h.now)
        XCTAssertFalse(state.isDirty)
        XCTAssertTrue(h.notifier.sent.isEmpty)
    }

    func testRestoreFailurePropagatesWithoutTouchingTheFile() throws {
        let h = AppStateHarness()
        defer { h.dispose() }
        let state = h.create()
        let bad = h.dir.file("bad.json")
        try Data("{not json".utf8).write(to: bad)
        let before = try Data(contentsOf: h.claudeConfigURL)
        XCTAssertThrowsError(try state.restoreClaudeConfig(from: bad)) { error in
            XCTAssertEqual(error as? ClaudeConfigError, .malformed("backup bad.json is not a valid config file"))
        }
        XCTAssertEqual(try Data(contentsOf: h.claudeConfigURL), before)
    }

    func testDisposeStopsTheWatchers() throws {
        let h = AppStateHarness()
        defer { h.dispose() }
        let state = h.create()
        Thread.sleep(forTimeInterval: 0.3)
        state.dispose()
        try h.writeClaudeServers([("scoutbook", try XCTUnwrap(state.store.mcps["scoutbook"]).config)])
        XCTAssertFalse(h.ui.pumpUntil({ !h.notifier.sent.isEmpty }, timeout: settle))
        XCTAssertEqual(try h.claudeServers().keys.sorted(), ["scoutbook"])   // nobody regenerated it
    }
}
